File: snex/src/commands/namestyle.rs
use std::collections::HashMap;
use std::sync::Mutex;

use once_cell::sync::Lazy;
use serde_json::json;
use serenity::builder::{
    CreateActionRow, CreateInputText, CreateInteractionResponse,
    CreateInteractionResponseMessage, CreateModal, CreateSelectMenu, CreateSelectMenuKind,
    CreateSelectMenuOption,
};
use serenity::http::Http;
use serenity::model::application::{ComponentInteraction, ModalInteraction};
use serenity::model::id::GuildId;
use serenity::model::input_text::InputTextStyle;

pub const FONTS: &[(u8, &str)] = &[
    (1, "Bangers"),
    (2, "BioRhyme"),
    (3, "Sakura"),
    (4, "Chicle"),
    (5, "Compagnon"),
    (6, "MuseoModerno"),
    (7, "Neo-Castel"),
    (8, "Pixelify Sans"),
    (9, "Ribes"),
    (10, "Sinistre"),
    (11, "Default (GG Sans)"),
    (12, "Zilla Slab"),
];

pub const EFFECTS: &[(u8, &str)] = &[
    (1, "Solid"),
    (2, "Gradient"),
    (3, "Neon"),
    (4, "Toon"),
    (5, "Pop"),
    (6, "Shadow"),
];

fn font_name(id: u8) -> &'static str {
    FONTS.iter().find(|(k, _)| *k == id).map(|(_, v)| *v).unwrap_or("?")
}

fn effect_name(id: u8) -> &'static str {
    EFFECTS.iter().find(|(k, _)| *k == id).map(|(_, v)| *v).unwrap_or("?")
}

#[derive(Default, Clone, Copy)]
struct PendingStyle {
    font_id: Option<u8>,
    effect_id: Option<u8>,
}

static PENDING_STYLE: Lazy<Mutex<HashMap<u64, PendingStyle>>> =
    Lazy::new(|| Mutex::new(HashMap::new()));

fn hex_to_int(hex: &str) -> anyhow::Result<u32> {
    let cleaned = hex.trim().trim_start_matches('#');
    u32::from_str_radix(cleaned, 16).map_err(|_| anyhow::anyhow!("не hex-цвет"))
}

async fn apply_name_style(
    http: &Http,
    guild_id: GuildId,
    font_id: u8,
    effect_id: u8,
    colors: Vec<u32>,
) -> anyhow::Result<()> {
    let route = format!("/guilds/{guild_id}/members/@me");
    let body = json!({
        "display_name_font_id": font_id,
        "display_name_effect_id": effect_id,
        "display_name_colors": colors,
    });
    http.client()
        .patch(format!("https://discord.com/api/v10{route}"))
        .header("Authorization", format!("Bot {}", http.token()))
        .json(&body)
        .send()
        .await?
        .error_for_status()?;
    Ok(())
}

async fn reset_name_style(http: &Http, guild_id: GuildId) -> anyhow::Result<()> {
    let route = format!("/guilds/{guild_id}/members/@me");
    let body = json!({
        "display_name_font_id": null,
        "display_name_effect_id": null,
        "display_name_colors": null,
    });
    http.client()
        .patch(format!("https://discord.com/api/v10{route}"))
        .header("Authorization", format!("Bot {}", http.token()))
        .json(&body)
        .send()
        .await?
        .error_for_status()?;
    Ok(())
}

/// Строит компоненты панели /namestyle: два select-меню (шрифт, эффект)
/// и кнопку сброса.
pub fn build_namestyle_panel() -> Vec<CreateActionRow> {
    let font_options: Vec<CreateSelectMenuOption> = FONTS
        .iter()
        .map(|(id, name)| CreateSelectMenuOption::new(*name, id.to_string()))
        .collect();

    let effect_options: Vec<CreateSelectMenuOption> = EFFECTS
        .iter()
        .map(|(id, name)| CreateSelectMenuOption::new(*name, id.to_string()))
        .collect();

    let font_select = CreateSelectMenu::new(
        "namestyle_font_select",
        CreateSelectMenuKind::String { options: font_options },
    )
    .placeholder("Выбери шрифт");

    let effect_select = CreateSelectMenu::new(
        "namestyle_effect_select",
        CreateSelectMenuKind::String { options: effect_options },
    )
    .placeholder("Выбери эффект");

    let reset_button = serenity::builder::CreateButton::new("namestyle_reset_btn")
        .label("Сбросить стиль")
        .style(serenity::model::application::ButtonStyle::Danger);

    vec![
        CreateActionRow::SelectMenu(font_select),
        CreateActionRow::SelectMenu(effect_select),
        CreateActionRow::Buttons(vec![reset_button]),
    ]
}

/// Обрабатывает select-меню и кнопки панели /namestyle.
/// Возвращает true, если взаимодействие было обработано этим модулем.
pub async fn handle_namestyle_component(
    http: &Http,
    interaction: &ComponentInteraction,
) -> anyhow::Result<bool> {
    let Some(guild_id) = interaction.guild_id else {
        return Ok(false);
    };
    let custom_id = interaction.data.custom_id.as_str();

    match custom_id {
        "namestyle_font_select" => {
            let serenity::model::application::ComponentInteractionDataKind::StringSelect { values } =
                &interaction.data.kind
            else {
                return Ok(false);
            };
            let font_id: u8 = values.first().and_then(|v| v.parse().ok()).unwrap_or(11);

            {
                let mut state = PENDING_STYLE.lock().unwrap();
                let entry = state.entry(guild_id.get()).or_default();
                entry.font_id = Some(font_id);
            }

            let response = CreateInteractionResponseMessage::new()
                .content(format!(
                    "### Шрифт выбран: {}\n-# Теперь выбери эффект",
                    font_name(font_id)
                ))
                .ephemeral(true);
            interaction
                .create_response(http, CreateInteractionResponse::Message(response))
                .await?;
            Ok(true)
        }

        "namestyle_effect_select" => {
            let serenity::model::application::ComponentInteractionDataKind::StringSelect { values } =
                &interaction.data.kind
            else {
                return Ok(false);
            };
            let effect_id: u8 = values.first().and_then(|v| v.parse().ok()).unwrap_or(1);

            let font_id = {
                let mut state = PENDING_STYLE.lock().unwrap();
                let entry = state.entry(guild_id.get()).or_default();
                entry.effect_id = Some(effect_id);
                entry.font_id
            };

            let Some(font_id) = font_id else {
                let response = CreateInteractionResponseMessage::new()
                    .content("### Сначала выбери шрифт в select-меню выше")
                    .ephemeral(true);
                interaction
                    .create_response(http, CreateInteractionResponse::Message(response))
                    .await?;
                return Ok(true);
            };

            let needs_two = effect_id == 2; // Gradient
            let mut inputs = vec![CreateInputText::new(
                InputTextStyle::Short,
                "Цвет 1 (#hex)",
                "color1",
            )
            .placeholder("#FF69B4")
            .required(true)
            .max_length(7)];

            if needs_two {
                inputs.push(
                    CreateInputText::new(InputTextStyle::Short, "Цвет 2 (#hex) — для Gradient", "color2")
                        .placeholder("#5865F2")
                        .required(true)
                        .max_length(7),
                );
            }

            let modal = CreateModal::new(
                format!("namestyle_color_modal_{font_id}_{effect_id}"),
                "Цвет стиля ника",
            )
            .components(
                inputs
                    .into_iter()
                    .map(CreateActionRow::InputText)
                    .collect(),
            );

            interaction
                .create_response(http, CreateInteractionResponse::Modal(modal))
                .await?;
            Ok(true)
        }

        "namestyle_reset_btn" => {
            interaction.defer_ephemeral(http).await?;
            match reset_name_style(http, guild_id).await {
                Ok(()) => {
                    PENDING_STYLE.lock().unwrap().remove(&guild_id.get());
                    interaction
                        .create_followup(
                            http,
                            serenity::builder::CreateInteractionResponseFollowup::new()
                                .content("### Стиль сброшен")
                                .ephemeral(true),
                        )
                        .await?;
                }
                Err(err) => {
                    interaction
                        .create_followup(
                            http,
                            serenity::builder::CreateInteractionResponseFollowup::new()
                                .content(format!("`ошибка: {err}`"))
                                .ephemeral(true),
                        )
                        .await?;
                }
            }
            Ok(true)
        }

        _ => Ok(false),
    }
}

/// Обрабатывает отправку модалки с цветом(-ами).
pub async fn handle_namestyle_modal(
    http: &Http,
    interaction: &ModalInteraction,
) -> anyhow::Result<bool> {
    let Some(guild_id) = interaction.guild_id else {
        return Ok(false);
    };
    let custom_id = interaction.data.custom_id.as_str();

    if !custom_id.starts_with("namestyle_color_modal_") {
        return Ok(false);
    }

    let parts: Vec<&str> = custom_id
        .trim_start_matches("namestyle_color_modal_")
        .split('_')
        .collect();
    let (Some(font_str), Some(effect_str)) = (parts.first(), parts.get(1)) else {
        return Ok(true);
    };
    let font_id: u8 = font_str.parse().unwrap_or(11);
    let effect_id: u8 = effect_str.parse().unwrap_or(1);

    let mut color1_raw = String::new();
    let mut color2_raw = String::new();

    for row in &interaction.data.components {
        for component in &row.components {
            if let serenity::model::application::ActionRowComponent::InputText(input) = component {
                if let Some(value) = &input.value {
                    match input.custom_id.as_str() {
                        "color1" => color1_raw = value.clone(),
                        "color2" => color2_raw = value.clone(),
                        _ => {}
                    }
                }
            }
        }
    }

    let mut colors = Vec::new();
    match hex_to_int(&color1_raw) {
        Ok(c) => colors.push(c),
        Err(_) => {
            let response = CreateInteractionResponseMessage::new()
                .content("### Неверный формат цвета, используй `#rrggbb`")
                .ephemeral(true);
            interaction
                .create_response(http, CreateInteractionResponse::Message(response))
                .await?;
            return Ok(true);
        }
    }
    if !color2_raw.trim().is_empty() {
        match hex_to_int(&color2_raw) {
            Ok(c) => colors.push(c),
            Err(_) => {
                let response = CreateInteractionResponseMessage::new()
                    .content("### Неверный формат цвета, используй `#rrggbb`")
                    .ephemeral(true);
                interaction
                    .create_response(http, CreateInteractionResponse::Message(response))
                    .await?;
                return Ok(true);
            }
        }
    }

    if effect_id == 2 && colors.len() != 2 {
        let response = CreateInteractionResponseMessage::new()
            .content("### Для эффекта Gradient нужно указать 2 цвета")
            .ephemeral(true);
        interaction
            .create_response(http, CreateInteractionResponse::Message(response))
            .await?;
        return Ok(true);
    }

    interaction.defer_ephemeral(http).await?;

    match apply_name_style(http, guild_id, font_id, effect_id, colors).await {
        Ok(()) => {
            PENDING_STYLE.lock().unwrap().remove(&guild_id.get());
            let text = format!(
                "### Стиль применён\n> **Шрифт:** {}\n> **Эффект:** {}\n> **Цвет:** {}{}",
                font_name(font_id),
                effect_name(effect_id),
                color1_raw,
                if color2_raw.trim().is_empty() {
                    String::new()
                } else {
                    format!(" → {color2_raw}")
                }
            );
            interaction
                .create_followup(
                    http,
                    serenity::builder::CreateInteractionResponseFollowup::new()
                        .content(text)
                        .ephemeral(true),
                )
                .await?;
        }
        Err(err) => {
            interaction
                .create_followup(
                    http,
                    serenity::builder::CreateInteractionResponseFollowup::new()
                        .content(format!("`ошибка: {err}`"))
                        .ephemeral(true),
                )
                .await?;
        }
    }

    Ok(true)
}

File: snex/src/NameStyle.cs
// Команда /namestyle — панель управления стилем ника бота на сервере
// (шрифт, эффект, цвет/градиент), через недокументированный эндпоинт
// Discord PATCH /guilds/{guild_id}/members/@me.
//
// Стиль хранится на стороне Discord, а не у нас — переживает рестарт
// и офлайн бота, поэтому базы данных тут не нужно.

using System.Collections.Concurrent;
using System.Net.Http.Json;
using NetCord;
using NetCord.Rest;
using NetCord.Services.ApplicationCommands;
using NetCord.Services.ComponentInteractions;

namespace Snex;

public static class NameStyleData
{
    public static readonly (byte Id, string Name)[] Fonts =
    [
        (1, "Bangers"),
        (2, "BioRhyme"),
        (3, "Sakura"),
        (4, "Chicle"),
        (5, "Compagnon"),
        (6, "MuseoModerno"),
        (7, "Neo-Castel"),
        (8, "Pixelify Sans"),
        (9, "Ribes"),
        (10, "Sinistre"),
        (11, "Default (GG Sans)"),
        (12, "Zilla Slab"),
    ];

    public static readonly (byte Id, string Name)[] Effects =
    [
        (1, "Solid"),
        (2, "Gradient"),
        (3, "Neon"),
        (4, "Toon"),
        (5, "Pop"),
        (6, "Shadow"),
    ];

    public static string FontName(byte id) => Fonts.FirstOrDefault(f => f.Id == id).Name ?? "?";
    public static string EffectName(byte id) => Effects.FirstOrDefault(e => e.Id == id).Name ?? "?";

    // Временное состояние выбора шрифта/эффекта по guild_id, пока
    // пользователь не введёт цвет(а) в модалке. Хранится в памяти —
    // переживать рестарт не нужно, это промежуточный шаг одной
    // настройки за один присест.
    public static readonly ConcurrentDictionary<ulong, (byte? FontId, byte? EffectId)> Pending = new();

    public static int HexToInt(string hex)
    {
        var cleaned = hex.Trim().TrimStart('#');
        return Convert.ToInt32(cleaned, 16);
    }

    public static async Task ApplyAsync(RestClient rest, ulong guildId, byte fontId, byte effectId, int[] colors)
    {
        var body = new
        {
            display_name_font_id = fontId,
            display_name_effect_id = effectId,
            display_name_colors = colors,
        };
        await PatchSelfMemberAsync(rest, guildId, body);
    }

    public static async Task ResetAsync(RestClient rest, ulong guildId)
    {
        var body = new
        {
            display_name_font_id = (byte?)null,
            display_name_effect_id = (byte?)null,
            display_name_colors = (int[]?)null,
        };
        await PatchSelfMemberAsync(rest, guildId, body);
    }

    private static async Task PatchSelfMemberAsync(RestClient rest, ulong guildId, object body)
    {
        using var http = new HttpClient();
        http.DefaultRequestHeaders.Authorization =
            new System.Net.Http.Headers.AuthenticationHeaderValue("Bot", rest.Token?.RawToken ?? "");

        var response = await http.PatchAsJsonAsync(
            $"https://discord.com/api/v10/guilds/{guildId}/members/@me", body);
        response.EnsureSuccessStatusCode();
    }

    /// Собирает Components V2 сообщение с одним текстовым блоком —
    /// используется для всех служебных ответов namestyle (статусы, ошибки).
    public static InteractionMessageProperties TextV2(string text)
    {
        var container = new ComponentContainerProperties
        {
            new TextDisplayProperties(text),
        };

        return new InteractionMessageProperties
        {
            Flags = MessageFlags.Ephemeral | MessageFlags.IsComponentsV2,
            Components = [container],
        };
    }
}

public class NameStyleCommandModule : ApplicationCommandModule<ApplicationCommandContext>
{
    [SlashCommand("namestyle", "Настроить стиль ника бота на этом сервере")]
    public async Task NamestyleAsync()
    {
        var fontMenu = new StringMenuProperties("namestyle_font_select")
            .WithPlaceholder("Выбери шрифт")
            .AddOptions(NameStyleData.Fonts
                .Select(f => new StringMenuSelectOptionProperties(f.Name, f.Id.ToString())));

        var effectMenu = new StringMenuProperties("namestyle_effect_select")
            .WithPlaceholder("Выбери эффект")
            .AddOptions(NameStyleData.Effects
                .Select(e => new StringMenuSelectOptionProperties(e.Name, e.Id.ToString())));

        var resetButton = new ButtonProperties("namestyle_reset_btn", "Сбросить стиль", ButtonStyle.Danger);

        var container = new ComponentContainerProperties
        {
            new TextDisplayProperties("### Стиль ника бота\n-# Выбери шрифт и эффект, затем укажи цвет(а)"),
            fontMenu,
            effectMenu,
            new ActionRowProperties().AddComponents(resetButton),
        };

        await RespondAsync(InteractionCallback.Message(new InteractionMessageProperties
        {
            Flags = MessageFlags.Ephemeral | MessageFlags.IsComponentsV2,
            Components = [container],
        }));
    }
}

public class NameStyleComponentModule : ComponentInteractionModule<StringMenuInteractionContext>
{
    [ComponentInteraction("namestyle_font_select")]
    public async Task FontSelectAsync()
    {
        var guildId = Context.Interaction.GuildId!.Value;
        var fontId = byte.Parse(Context.SelectedValues[0]);

        NameStyleData.Pending.AddOrUpdate(
            guildId,
            (fontId, null),
            (_, existing) => (fontId, existing.EffectId));

        await RespondAsync(InteractionCallback.Message(
            NameStyleData.TextV2($"### Шрифт выбран: {NameStyleData.FontName(fontId)}\n-# Теперь выбери эффект")));
    }

    [ComponentInteraction("namestyle_effect_select")]
    public async Task EffectSelectAsync()
    {
        var guildId = Context.Interaction.GuildId!.Value;
        var effectId = byte.Parse(Context.SelectedValues[0]);

        var state = NameStyleData.Pending.AddOrUpdate(
            guildId,
            (null, effectId),
            (_, existing) => (existing.FontId, effectId));

        if (state.FontId is not byte fontId)
        {
            await RespondAsync(InteractionCallback.Message(
                NameStyleData.TextV2("### Сначала выбери шрифт в select-меню выше")));
            return;
        }

        var needsTwo = effectId == 2; // Gradient
        var color1Input = new LabelProperties(
            "Цвет 1 (#hex)",
            new TextInputProperties("color1", TextInputStyle.Short)
                .WithPlaceholder("#FF69B4")
                .WithRequired(true)
                .WithMaxLength(7));

        var modal = new ModalProperties("namestyle_color_modal", "Цвет стиля ника")
            .AddComponents(color1Input);

        if (needsTwo)
        {
            var color2Input = new LabelProperties(
                "Цвет 2 (#hex) — для Gradient",
                new TextInputProperties("color2", TextInputStyle.Short)
                    .WithPlaceholder("#5865F2")
                    .WithRequired(true)
                    .WithMaxLength(7));
            modal.AddComponents(color2Input);
        }

        await RespondAsync(InteractionCallback.Modal(modal));
    }
}

public class NameStyleButtonModule : ComponentInteractionModule<ButtonInteractionContext>
{
    [ComponentInteraction("namestyle_reset_btn")]
    public async Task ResetAsync()
    {
        var guildId = Context.Interaction.GuildId!.Value;

        await RespondAsync(InteractionCallback.DeferredModifyMessage);

        try
        {
            await NameStyleData.ResetAsync(Context.Client.Rest, guildId);
            NameStyleData.Pending.TryRemove(guildId, out _);
            await FollowupAsync(NameStyleData.TextV2("### Стиль сброшен"));
        }
        catch (Exception ex)
        {
            await FollowupAsync(NameStyleData.TextV2($"`ошибка: {ex.Message}`"));
        }
    }
}

public class NameStyleModalModule : ComponentInteractionModule<ModalInteractionContext>
{
    [ComponentInteraction("namestyle_color_modal")]
    public async Task ColorModalAsync()
    {
        var guildId = Context.Interaction.GuildId!.Value;

        if (!NameStyleData.Pending.TryGetValue(guildId, out var state)
            || state.FontId is not byte fontId
            || state.EffectId is not byte effectId)
        {
            await RespondAsync(InteractionCallback.Message(
                NameStyleData.TextV2("### Сессия настройки истекла, начни заново с `/namestyle`")));
            return;
        }

        var components = Context.Interaction.Data.Components;
        var color1Raw = components.OfType<TextInput>().FirstOrDefault(t => t.CustomId == "color1")?.Value ?? "";
        var color2Raw = components.OfType<TextInput>().FirstOrDefault(t => t.CustomId == "color2")?.Value ?? "";

        var colors = new List<int>();
        try
        {
            colors.Add(NameStyleData.HexToInt(color1Raw));
        }
        catch
        {
            await RespondAsync(InteractionCallback.Message(
                NameStyleData.TextV2("### Неверный формат цвета, используй `#rrggbb`")));
            return;
        }

        if (!string.IsNullOrWhiteSpace(color2Raw))
        {
            try
            {
                colors.Add(NameStyleData.HexToInt(color2Raw));
            }
            catch
            {
                await RespondAsync(InteractionCallback.Message(
                    NameStyleData.TextV2("### Неверный формат цвета, используй `#rrggbb`")));
                return;
            }
        }

        if (effectId == 2 && colors.Count != 2)
        {
            await RespondAsync(InteractionCallback.Message(
                NameStyleData.TextV2("### Для эффекта Gradient нужно указать 2 цвета")));
            return;
        }

        await RespondAsync(InteractionCallback.DeferredModifyMessage);

        try
        {
            await NameStyleData.ApplyAsync(Context.Client.Rest, guildId, fontId, effectId, colors.ToArray());
            NameStyleData.Pending.TryRemove(guildId, out _);

            var colorText = colors.Count > 1 ? $"{color1Raw} → {color2Raw}" : color1Raw;
            var text = $"### Стиль применён\n" +
                       $"> **Шрифт:** {NameStyleData.FontName(fontId)}\n" +
                       $"> **Эффект:** {NameStyleData.EffectName(effectId)}\n" +
                       $"> **Цвет:** {colorText}";

            await FollowupAsync(NameStyleData.TextV2(text));
        }
        catch (Exception ex)
        {
            await FollowupAsync(NameStyleData.TextV2($"`ошибка: {ex.Message}`"));
        }
    }
}